src/middlewared/middlewared/plugins/jbof/crud.py

```diff
@@ -0,0 +1,546 @@
+                        'mtu': static_mtu(),
+                        'check': {'ping_ip': shelf_ip,
+                                  'ping_mac': shelf_ip_to_mac[shelf_ip]}
+                    }
+                    if node:
+                        payload['node'] = node
+                    if await self.middleware.call('rdma.interface.create', payload):
+                        dirty = True
+                        connected_shelf_ips.add(shelf_ip)
+                        # break out of the ifname loop
+                        if node:
+                            self.logger.info(f'Created link on node {node}: {host_ip} -> {shelf_ip}')
+                        else:
+                            self.logger.info(f'Created link: {host_ip} -> {shelf_ip}')
+                        break
+        return list(connected_shelf_ips)
+
+    @private
+    async def attach_drives(self, mgmt_ip, shelf_index, schema, verrors):
+        """Attach drives from the specified expansion shelf."""
+        if await self.middleware.call('failover.licensed'):
+            # HA system
+            if not await self.middleware.call('failover.remote_connected'):
+                verrors.add(schema, 'HA system must be in good state to allow configuration.')
+                return
+
+            this_node = await self.middleware.call('failover.node')
+            if this_node not in ['A', 'B']:
+                verrors.add(schema, 'HA system must be in good state to allow configuration: Invalid node')
+                return
+
+            for node in ['A', 'B']:
+                await self.attach_drives_to_node(node, shelf_index)
+        else:
+            await self.attach_drives_to_node(node, shelf_index)
+
+    @private
+    async def attach_drives_to_node(self, node, shelf_index):
+        localnode = not node or node == await self.middleware.call('failover.node')
+        configured_interfaces = await self.middleware.call('rdma.interface.query')
+        if localnode:
+            for interface in configured_interfaces:
+                if interface['node'] != node:
+                    continue
+                jbof_ip = jbof_static_ip_from_initiator_ip(interface['address'])
+                await self.middleware.call('jbof.nvme_connect', jbof_ip)
+        else:
+            for interface in configured_interfaces:
+                if interface['node'] != node:
+                    continue
+                jbof_ip = jbof_static_ip_from_initiator_ip(interface['address'])
+                await self.middleware.call('failover.call_remote', 'jbof.nvme_connect', [jbof_ip])
+
+    @private
+    def nvme_connect(self, ip, nr_io_queues=16):
+        command = ['nvme', 'connect-all', '-t', 'rdma', '-a', ip, '--persistent', '-i', f'{nr_io_queues}']
+        ret = subprocess.run(command, capture_output=True)
+        if ret.returncode:
+            self.logger.debug(f'Failed to execute "{" ".join(command)}": {ret.stderr.decode()}')
```

Comment:
We need to check for stderr and/or stdout (unless you're 100% positive nvme binary always reports error to stderr). We also need to decode the stderr once. (i.e.
```
if ret.returncode:
    error = ret.stderr.decode() if stderr else ret.stdout.decode()
    if not error:
        error = 'No error message reported'
```
We also need to use `%s` operators in the logger calls for better efficiency i.e
```
self.logger.debug('Failed to execute command: %r with error: %r', " ".join(command), error)